README
Jconf is a library that permits to manage the file-based configuration of programs using json files.

build:
-----

 autoreconf -ivf
 ./configure
 make
 make install ( or sudo make install)

C++ Wrapper:
------------

A C++ 11 wrapper of this library is included  (CppWrapper directory).
To build the C++ library that include the wrapper, the following steps are required:

cd CppWrapper
autoreconf -ivf
./configure
 make
make install ( or sudo make install)


CppWrapper/ChangeLog
JsonLibCppWrapper (0.0.1) any; urgency=low 

 * Beta Test
 Initial Version.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com> Wed, 01 Jan 2018 02:59:00 -0100 


test/listAddresses.c
// -----------------------------------------------------------------
// Libjparse - A JSON configuration files parser.
// Copyright (C) 2014  Gabriele Bonacini
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
// -----------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <parseJsonConfig.h>

int printNodes(node* nodeElement, int node);

int main(int argc, char** argv){

  node nodeRoot;
  node* nodeCurrent;
  int result=0, node=0;

  pathIndex* fullIndexesList;

 if(argc < 2){
		fprintf(stderr,"Test file path needed!");
		exit(1);
 }
 
 result=parseJsonConfig(argv[1], &nodeRoot, &fullIndexesList);

 if(result<=0){
	fprintf(stderr,"Parse status: NOK Err: %d\n", result);
	exit(-1);
 }

 nodeCurrent=&nodeRoot;

 while(nodeCurrent!=NULL){
		node++;
		if(printNodes(nodeCurrent, node) == 0){
			if(nodeCurrent->innerNode!=NULL)
				nodeCurrent=nodeCurrent->innerNode;
			else if(nodeCurrent->nextNode!=NULL)
				nodeCurrent=nodeCurrent->nextNode;
			else
				nodeCurrent=nodeCurrent->exitNode;
		}else{
			fprintf(stderr,"Print status: NOK Err: %d\n", result);
			exit(-1);
		}

 }

 return 0;
}

int printNodes(node* nodeElement, int node){
		int err=0;
		if(nodeElement != NULL){
			fprintf(stderr,"NODE;%d;Label;%s;",node ,nodeElement->nodeName);
			switch(nodeElement->nodeType){
				case ROOT_T:
					fprintf(stderr,"Type;ROOT_T;Value;0;");
				break;
				case EMPTY_T:
					fprintf(stderr,"Type;EMPTY_T;Value;0;");
				break;
				case  NUMERIC_T:
					fprintf(stderr,"Type;NUMERIC_T;Value;%f;",(nodeElement->data).number);
				break;
				case  TEXT_T:
					fprintf(stderr,"Type;TEXT_T;Value;%s;",(nodeElement->data).string);
				break;
				case  BOOL_T:
					fprintf(stderr,"Type;BOOL_T;Value;%d;",(nodeElement->data).boolean);
				break;
				case  MATRIX_T:
					fprintf(stderr,"Type;MATRIX_T;Value;0;");
				break;
				case  OBJ_T:
					fprintf(stderr,"Type;OBJ_T;Value;0;");
				break;
				case UNASSIGNED_T:
					fprintf(stderr,"Type;UNASSIGNED;Type;%d;",nodeElement->nodeType);
				break;
			}
			fprintf(stderr,"node;%p;previousNode;%p;nextNode;%p;innerNode;%p;outerNode;%p;exitNode;%p\n",\
				nodeElement, nodeElement->previousNode, nodeElement->nextNode,\
				nodeElement->innerNode, nodeElement->outerNode, nodeElement->exitNode );
		}else{
			fprintf(stderr,"NODE -- Null node !");
			err=-1;
		}
		return err;
}


test/parseJsonFile.c
// -----------------------------------------------------------------
// Libjparse - A JSON configuration files parser.
// Copyright (C) 2014  Gabriele Bonacini
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
// -----------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>

#include <parseJsonConfig.h>

int printNodeValue(node* nodeStack);
__attribute__((noreturn)) void printErrorMesg(char* msg);

static bool checktype=false, debug=false;

int main(int argc, char** argv){

  char jsonFile[MAX_NAME_SIZE];
  char url[MAX_NAME_SIZE];

  struct stat statBuffer;

  bool fileName=false, value=false;
  node nodeRoot;
  node* tempNode;
  int result=0, c=0;

  pathIndex* fullIndexesList;
  pathIndex* element;

  while ( (c = getopt(argc, argv, "df:v:ht")) != EOF) {
	switch (c) {
	case 'd':               /* Debug switch  */
		debug=true;
	break;
	case 'f':               /* Json File  */
		fileName=true;
		memset(jsonFile,0,MAX_NAME_SIZE);
		strncpy(jsonFile, optarg, MAX_NAME_SIZE-1);
		if(stat(jsonFile, &statBuffer)){
			fprintf(ERRSTREAM, "The file %s do not exist!", jsonFile);
			exit (1);
		}
	break;
	case 'h':               /* Help */
			printErrorMesg(NULL);
			// The program exits
	case 't':               /* Type */
			checktype=true;
	break;
	case 'v':               /* Value */
		value=true;
		memset(url,0,MAX_NAME_SIZE);
		strncpy(url, optarg, MAX_NAME_SIZE-1);
	break;
	case '?':		/* Default */
			printErrorMesg(NULL);

	}
 }

 if(!fileName){
		printErrorMesg("The json file was not specified.!\n");
 }
 if(!value){
		printErrorMesg("The url was not specified.!\n");
 }
 
 result=parseJsonConfig((char const *)jsonFile, &nodeRoot, &fullIndexesList);

 if(result>0){
	verboseLong(debug, "Parse status: OK IndexLen: %d\n", result);
 }else{
	verboseLong(debug, "Parse status: NOK Err: %d\n", result);
	exit(-1);
 }



 element=getElementValueByString(fullIndexesList, (char const *)url); 
 verbose(debug, "------------- End Search ---------------\n");
 if(element!=NULL){
	tempNode=element->Node;
	printNodeValue(tempNode);
 }else{
	 fprintf(ERRSTREAM,"Element: Not Found\n");
 }

 verbose(debug, "------------- End Result  ---------------\n");

 return 0;
}

void printErrorMesg(char* msg){
	fprintf(ERRSTREAM, "%s", msg);
	if(msg!=NULL)
	fprintf(ERRSTREAM, "%s%s%s%s%s","Usage:\n -f<Json File> \t\t To set the Json file to process.\n",\
					" -v<Value> \t\t To set the value to search.\n",\
					" -d \t\t To print verbose tracing messages.\n",\
					" -t \t\t To print the type of the value.\n",\
					" -h \t\t To print this help message.\n");
	exit(1);
}

int printNodeValue(node* nodeElement){
			int err=0;
			node* currNode=NULL;

			if(nodeElement != NULL){
				verboseLong(debug,"NODE -- Label:%s -- ",nodeElement->nodeName);
				if(debug) 
					fprintf(ERRSTREAM,"node: %p - previousNode: %p - nextNode: %p - innerNode: %p - outerNode: %p\n",\
						nodeElement, nodeElement->previousNode, nodeElement->nextNode,\
						nodeElement->innerNode, nodeElement->outerNode );
				switch(nodeElement->nodeType){
					case ROOT_T:
						if(checktype)
							fprintf(ERRSTREAM," This is the ROOT node.\n");
					break;
					case EMPTY_T:
						if(checktype)
							fprintf(ERRSTREAM," EMPTY_T is present.\n");
					break;
					case  NUMERIC_T:
						if(checktype)
							fprintf(ERRSTREAM," NUMERIC_T is present.\nValue: \n");
						fprintf(ERRSTREAM,"%f\n",(nodeElement->data).number);
					break;
					case  TEXT_T:
						if(checktype)
							fprintf(ERRSTREAM," TEXT_T is present.\nValue: \n");
						fprintf(ERRSTREAM,"%s\n",(nodeElement->data).string);
					break;
					case  BOOL_T:
						if(checktype)
							fprintf(ERRSTREAM," BOOL_T is present.\nValue: \n");
						fprintf(ERRSTREAM,"%d\n",(nodeElement->data).boolean);
					break;
					case  MATRIX_T:
						if(checktype)
							fprintf(ERRSTREAM," MATRIX_T is present.\n");
						currNode=nodeElement->innerNode;
						while(currNode!=NULL){
							if(currNode->innerNode!=NULL)
								currNode=currNode->innerNode;
							else if(currNode->nextNode!=NULL)
								currNode=currNode->nextNode;
							else
								currNode=NULL;
						}
					break;
					case  OBJ_T:
						fprintf(ERRSTREAM," OBJ_T is present \n ");
					break;
					case UNASSIGNED_T:
						fprintf(ERRSTREAM," Error: UNASSIGNED is present. ");
						fprintf(ERRSTREAM," Type:%d \n ",nodeElement->nodeType);
					break;
				}
                        }else{
				fprintf(ERRSTREAM,"NODE -- Null node !\n");
				err=-1;
			}
			return err;
}


test/parseJsonFileMatrix.c
// -----------------------------------------------------------------
// Libjparse - A JSON configuration files parser.
// Copyright (C) 2014  Gabriele Bonacini
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
// -----------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>

#include <parseJsonConfig.h>

int printNodeValue(node* nodeStack);
__attribute__((noreturn)) void printErrorMesg(char* msg);

static bool checktype=false, debug=false;

int main(int argc, char** argv){

  char jsonFile[MAX_NAME_SIZE];
  char url[MAX_NAME_SIZE];

  int index=0;

  struct stat statBuffer;

  bool fileName=false, value=false;
  node nodeRoot;
  node* tempNode;
  int result=0, c=0;

  pathIndex* fullIndexesList;
  pathIndex* element;

  while ( (c = getopt(argc, argv, "df:v:ht")) != EOF) {
	switch (c) {
	case 'd':               /* Debug switch  */
		debug=true;
	break;
	case 'f':               /* Json File  */
		fileName=true;
		memset(jsonFile,0,MAX_NAME_SIZE);
		strncpy(jsonFile, optarg, MAX_NAME_SIZE-1);
		if(stat(jsonFile, &statBuffer)){
			fprintf(ERRSTREAM, "The file %s do not exist!", jsonFile);
			exit (1);
		}
	break;
	case 'h':               /* Help */
			printErrorMesg(NULL); 
			// The program exits
	case 't':               /* Type */
			checktype=true;
	break;
	case 'v':               /* Value */
		value=true;
		memset(url,0,MAX_NAME_SIZE);
		strncpy(url, optarg, MAX_NAME_SIZE-1);
	break;
	case '?':		/* Default */
			printErrorMesg(NULL);

	}
 }

 if(!fileName){
		printErrorMesg("The json file was not specified.!\n");
 }
 if(!value){
		printErrorMesg("The url was not specified.!\n");
 }
 
 result=parseJsonConfig((char const *)jsonFile, &nodeRoot, &fullIndexesList);

 if(result>0){
	verboseLong(debug, "Parse status: OK IndexLen: %d\n", result);
 }else{
	verboseLong(debug, "Parse status: NOK Err: %d\n", result);
	exit(-1);
 }



 element=getElementValueByString(fullIndexesList, (char const *)url); 
 verbose(debug, "------------- End Search ---------------\n");

 tempNode=getArrayOfObjsElements(element->Node);
 if(tempNode!=NULL){
	tempNode=getElementFromObj(tempNode);
	fprintf(ERRSTREAM, "%d:", index);
	printNodeValue(tempNode);
	while((tempNode=getElementFromObj(NULL))){
		printNodeValue(tempNode);
	}
	fprintf(ERRSTREAM, "\n");
	index++;
 }

 while((tempNode=getArrayOfObjsElements(NULL))){
	 if(tempNode!=NULL){
		tempNode=getElementFromObj(tempNode);
		fprintf(ERRSTREAM, "%d:", index);
		printNodeValue(tempNode);
		while((tempNode=getElementFromObj(NULL))){
			printNodeValue(tempNode);
		}
		fprintf(ERRSTREAM, "\n");
		index++;
	 }
 }


 verbose(debug, "------------- End Result  ---------------\n");

 return 0;
}

void printErrorMesg(char* msg){
	fprintf(ERRSTREAM, "%s", msg);
	if(msg!=NULL)
	fprintf(ERRSTREAM, "%s%s%s%s%s","Usage:\n -f<Json File> \t\t To set the Json file to process.\n",\
					" -v<Value> \t\t To set the value to search.\n",\
					" -d \t\t To print verbose tracing messages.\n",\
					" -t \t\t To print the type of the value.\n",\
					" -h \t\t To print this help message.\n");
	exit(1);
}

int printNodeValue(node* nodeElement){
			int err=0;
			node* currNode=NULL;

			if(nodeElement != NULL){
				verboseLong(debug,"NODE -- Label:%s -- ",nodeElement->nodeName);
				if(debug) 
					fprintf(ERRSTREAM,"node: %p - previousNode: %p - nextNode: %p - innerNode: %p - outerNode: %p\n",\
						nodeElement, nodeElement->previousNode, nodeElement->nextNode,\
						nodeElement->innerNode, nodeElement->outerNode );
				switch(nodeElement->nodeType){
					case ROOT_T:
						if(checktype)
							fprintf(ERRSTREAM," This is the ROOT node.\nValue: \n");
					break;
					case EMPTY_T:
						if(checktype)
							fprintf(ERRSTREAM," EMPTY_T is present.\n");
							fprintf(ERRSTREAM,"<null>:");
					break;
					case  NUMERIC_T:
						if(checktype)
							fprintf(ERRSTREAM," NUMERIC_T is present.\nValue: \n");
						fprintf(ERRSTREAM,"%f:",(nodeElement->data).number);
					break;
					case  TEXT_T:
						if(checktype)
							fprintf(ERRSTREAM," TEXT_T is present.\nValue: \n");
						fprintf(ERRSTREAM,"%s:",(nodeElement->data).string);
					break;
					case  BOOL_T:
						if(checktype)
							fprintf(ERRSTREAM," BOOL_T is present.\nValue: \n");
						fprintf(ERRSTREAM,"%d:",(nodeElement->data).boolean);
					break;
					case  MATRIX_T:
						if(checktype)
							fprintf(ERRSTREAM," MATRIX_T is present.\n");
						currNode=nodeElement->innerNode;
						while(currNode!=NULL){
							if(currNode->innerNode!=NULL)
								currNode=currNode->innerNode;
							else if(currNode->nextNode!=NULL)
								currNode=currNode->nextNode;
							else
								currNode=NULL;
						}
					break;
					case  OBJ_T:
						fprintf(ERRSTREAM," OBJ_T is present \n ");
					break;
					case UNASSIGNED_T:
						fprintf(ERRSTREAM," Error: UNASSIGNED is present. ");
						fprintf(ERRSTREAM," Type:%d \n ",nodeElement->nodeType);
					break;
				}
                        }else{
				fprintf(ERRSTREAM,"<null>:");
				err=-1;
			}
			return err;
}


include/jsonGram.h
/* A Bison parser, made by GNU Bison 2.3.  */

/* Skeleton interface for Bison's Yacc-like parsers in C

   Copyright (C) 1984, 1989, 1990, 2000, 2001, 2002, 2003, 2004, 2005, 2006
   Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor,
   Boston, MA 02110-1301, USA.  */

/* As a special exception, you may create a larger work that contains
   part or all of the Bison parser skeleton and distribute that work
   under terms of your choice, so long as that work isn't itself a
   parser generator using the skeleton or a modified version thereof
   as a parser skeleton.  Alternatively, if you modify or redistribute
   the parser skeleton itself, you may (at your option) remove this
   special exception, which will cause the skeleton and the resulting
   Bison output files to be licensed under the GNU General Public
   License without this special exception.

   This special exception was added by the Free Software Foundation in
   version 2.2 of Bison.  */

/* Tokens.  */
#ifndef YYTOKENTYPE
# define YYTOKENTYPE
   /* Put the tokens into the symbol table, so that GDB and other debuggers
      know about them.  */
   enum yytokentype {
     STARTBLOCK = 258,
     ENDBLOCK = 259,
     STARTARRAY = 260,
     ENDARRAY = 261,
     TAG = 262,
     VOIDVALUE = 263,
     SEPARATOR = 264,
     BOOLEAN = 265,
     NUMBER = 266,
     STRING = 267
   };
#endif
/* Tokens.  */
#define STARTBLOCK 258
#define ENDBLOCK 259
#define STARTARRAY 260
#define ENDARRAY 261
#define TAG 262
#define VOIDVALUE 263
#define SEPARATOR 264
#define BOOLEAN 265
#define NUMBER 266
#define STRING 267




#if ! defined YYSTYPE && ! defined YYSTYPE_IS_DECLARED
typedef union YYSTYPE
#line 38 "/Users/bg/_Sorgenti/_Json/_Json_LY_hg/src/jsonGram.y"
{
        bool boolean;
        double number;
        char *string;
}
/* Line 1529 of yacc.c.  */
#line 79 "/Users/bg/_Sorgenti/_Json/_Json_LY_hg/src/../include/jsonGram.h"
	YYSTYPE;
# define yystype YYSTYPE /* obsolescent; will be withdrawn */
# define YYSTYPE_IS_DECLARED 1
# define YYSTYPE_IS_TRIVIAL 1
#endif

extern YYSTYPE yylval;



src/parseJsonConfig.c
// -----------------------------------------------------------------
// Libjparse - A JSON configuration files parser.
// Copyright (C) 2014  Gabriele Bonacini
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
// -----------------------------------------------------------------

#include <jsonParserTypes.h>
#include <parseJsonConfig.h>

#include <stdlib.h>
#include <string.h>
#include <stdio.h>

int parseJsonConfig(char const *configPath, node* nodeRoot, pathIndex** fullIndexesList){
	FILE* CONFIG= fopen(configPath, "r");
	if(CONFIG != NULL){
		yyin=CONFIG;
		yydebug=1;
		if(yyparse(nodeRoot)!= 0){
			return ERR_NO_PARSE;
		}
	}else{
		return ERR_NO_CONFIG_FILE;
	}
	(void)fclose(CONFIG);
	return createIndex(fullIndexesList);
}

int createIndex(pathIndex** fullIndexesList){
        int index=0, errstate=list;
        char buffer[MAX_INDEX_SIZE];
        pathBuffer* currentPath=NULL;
        pathBuffer* lastPath=NULL;
	node* lastPNode=NULL;

        #ifdef __DEBUG__JSON__
                fprintf(stderr,"Lex - Start  yywrap()\n");  
        #endif

        *fullIndexesList=(pathIndex*) calloc ((size_t)list, sizeof(pathIndex));
	if( fullIndexesList == NULL){
		memoryError();
	}

        for(index=0;index<list;index++){
                memset(buffer, 0, sizeof(buffer));
                currentPath=pathList[index];

                while(currentPath!=NULL ){
                        if( (MAX_INDEX_SIZE - ( 3 + strlen(buffer))) > strlen(*currentPath->nodePathComponent)){
                                strcat(buffer, *currentPath->nodePathComponent);
                                if(currentPath->Next!=NULL){
                                        strcat(buffer, ".");
                                }   
                        }else{
                                memset(buffer, 0, sizeof(buffer));
                                strcat(buffer, "String too long");
				errstate=-1;
                                break;
                        }   
			lastPath=currentPath;
			lastPNode=lastPath->Node;
                        currentPath=currentPath->Next;
			free(lastPath);
			lastPath=NULL;
                }   

                #ifdef __DEBUG__JSON__
                        fprintf(stderr,"Lex - Composed Path -- %s -- step -- %d\n", buffer, index);  
                #endif
		free(currentPath);
		currentPath=NULL;
                (*fullIndexesList + index)->fullNodeName=strdup(buffer);
                (*fullIndexesList + index)->Node=lastPNode; 

		pathList[index]=NULL;
        }   

        qsort (*fullIndexesList, (size_t)list, sizeof(pathIndex), indexesListCmp);

	return errstate;
}

int indexesListCmp (const void* sbuff1, const void* sbuff2){
  return strcmp (((pathIndex*)sbuff1)->fullNodeName, ((pathIndex*)sbuff2)->fullNodeName);
}

int indexesListSearch (const void* sbuff1, const void* sbuff2){
  #ifdef __DEBUG__JSON__
     fprintf(stderr,"Lex - Compare -- %s -- step -- %s\n", (char*)sbuff1, ((pathIndex*)sbuff2)->fullNodeName);  
  #endif
  return strcmp ((char*)sbuff1, ((pathIndex*)sbuff2)->fullNodeName);
}

/*@null@*/
pathIndex* getElementValueByString(pathIndex* fullIndexesList, char const *key){
	return bsearch( (void*)key, fullIndexesList, (size_t)list, sizeof (pathIndex), indexesListSearch);
}

bool checkNodeType(node* nodeElement, type expectType){
        if(nodeElement->nodeType == expectType)
                        return true;
        return false;
}

node* getArrayElements(node* arrayRootElement){
	static node* currentNode;
	errno=0;

	if(arrayRootElement == NULL){
		if(currentNode!=NULL){
			if(currentNode->nextNode!=NULL){
				currentNode=currentNode->nextNode;
				if(!isScalar(currentNode) || errno!=0 ){
				      errno=ERR_UNHANDLED_NODE_TYPE;
				      currentNode=NULL;
				}
			}else{
				errno=ERR_NO_MORE_ELEMENTS;
				currentNode=NULL;
			}
		}else{
			errno=ERR_WRONG_INITIAL_NODE;
		}
	}else{
		if(arrayRootElement->nodeType==MATRIX_T){
			currentNode=arrayRootElement->innerNode;
			if(!isScalar(currentNode) || errno!=0 ){
			      errno=ERR_UNHANDLED_NODE_TYPE;
			      currentNode=NULL;
			}
				
		}else{
			errno=ERR_WRONG_NODE_TYPE;
			currentNode=NULL;
		}
	} 
	
	return currentNode;
}


node* getArrayOfObjsElements(node* arrayRootElement){
	static node* currentNode;
	errno=0;

	if(arrayRootElement == NULL){
		if(currentNode!=NULL){
			if(currentNode->nextNode!=NULL){
				currentNode=currentNode->nextNode;
				if(currentNode->nodeType != OBJ_T){
				      errno=ERR_UNHANDLED_NODE_TYPE;
				      currentNode=NULL;
				}
			}else{
				errno=ERR_NO_MORE_ELEMENTS;
				currentNode=NULL;
			}
		}else{
			errno=ERR_WRONG_INITIAL_NODE;
		}
	}else{
		if(arrayRootElement->nodeType==MATRIX_T){
			currentNode=arrayRootElement->innerNode;
			if(currentNode==NULL){
				currentNode=arrayRootElement->nextNode;
			}
			if(currentNode->nodeType != OBJ_T){
			      errno=ERR_UNHANDLED_NODE_TYPE;
			      currentNode=NULL;
			}
				
		}else{
			errno=ERR_WRONG_NODE_TYPE;
			currentNode=NULL;
		}
	} 
	
	return currentNode;
}

node* getElementFromObj(node* object){
	static node* currentNode;
	errno=0;

	if(object == NULL){
		if(currentNode!=NULL){
			if(currentNode->nextNode!=NULL){
				currentNode=currentNode->nextNode;
				if(!isScalar(currentNode) || errno!=0){
				      errno=ERR_UNHANDLED_NODE_TYPE;
				      currentNode=NULL;
				}
			}else{
				errno=ERR_NO_MORE_ELEMENTS;
				currentNode=NULL;
			}
		}else{
			errno=ERR_WRONG_INITIAL_NODE;
		}
	}else{
		if(object->nodeType==OBJ_T){
			currentNode=object->innerNode;
			if(currentNode==NULL){
				currentNode=object->nextNode;
			}
			if(!isScalar(currentNode) || errno!=0){
			      errno=ERR_UNHANDLED_NODE_TYPE;
			      currentNode=NULL;
			}
				
		}else{
			errno=ERR_WRONG_NODE_TYPE;
			currentNode=NULL;
		}
	} 
	
	return currentNode;
}

bool isScalar(node* toCheck){
	bool result=false;
	switch(toCheck->nodeType){
		case OBJ_T:
		case MATRIX_T:
		break;
		case NUMERIC_T:
		case TEXT_T:
		case BOOL_T:
		case EMPTY_T:
			result=true;
		break;
		case ROOT_T:
		case UNASSIGNED_T:
			errno=ERR_WRONG_NODE_TYPE;
	}
	return result;
}



include/parseJsonConfig.h
// -----------------------------------------------------------------
// Libjparse - A JSON configuration files parser.
// Copyright (C) 2014  Gabriele Bonacini
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
// -----------------------------------------------------------------

#ifndef __PARSEJSONCONFIG__
#define __PARSEJSONCONFIG__

#include <stdio.h>

#include <jsonParserTypes.h>

int parseJsonConfig(char const *configPath, node* nodeRoot, pathIndex** fullIndexesList);

extern FILE * yyin;
extern int yyparse(node* nodeZero);
extern pathBuffer* pathList[MAX_NODE_STACK];

extern int yydebug;
extern int id, level, list;
extern node* currentNode;
extern node* nodeStack[MAX_NODE_STACK];
extern pathBuffer* pathList[MAX_NODE_STACK];

#endif



CppWrapper/include/jsonCppWrap.hpp
#ifndef __JSONCPPWRAP__
#define __JSONCPPWRAP__

#include <iostream>
#include <string>
#include <vector> 

extern "C" {
#include <jsonParserTypes.h>
#include <parseJsonConfig.h>
}

namespace jsonCppWrap{

class JsonCppWrapException{
	public:
		JsonCppWrapException(int errCode);
		JsonCppWrapException(std::string errMsg);
		JsonCppWrapException(std::string* errMsg);
		JsonCppWrapException(char* errMsg);
		JsonCppWrapException(const char* errMsg);
		JsonCppWrapException(int errCode, char* errMsg);
		JsonCppWrapException(int errCode, const char* errMsg);
		JsonCppWrapException(int errCode, std::string errMsg);
		JsonCppWrapException(int errCode, std::string* errMsg);
		std::string what(void) const noexcept;
	private:
		int errorCode;
		std::string errorMessage;
};

class JsonCppWrap{
	public:
		explicit JsonCppWrap(std::string* jsonFile) ;
		explicit JsonCppWrap(std::string& jsonFile) ;
		explicit JsonCppWrap(const char* jsonFile) ;
		node* getElementValueByPath(const char* elUrl) const noexcept;
		node* getElementValueByPath(std::string* elUrl) const noexcept;
		node* getElementValueByPath(std::string& elUrl) const noexcept;
		node* getElementValueByPathExc(const char* elUrl) const;
		node* getElementValueByPathExc(std::string* elUrl) const;
		node* getElementValueByPathExc(std::string& elUrl) const;
		type getType(node* nodeElement) const noexcept;
		node* getArrayElements(node* arrayRootElement) const noexcept;
		node* getArrayOfObjsElements(node* arrayRootElement) const noexcept;
		node* getElementFromObj(node* object) const noexcept;
		bool checkPath(std::string* elUrl) const noexcept;
		bool checkPath(std::string& elUrl) const noexcept;
		bool checkPath(char const* elUrl) const noexcept;
		bool checkPathList(std::vector<std::string>& urlList) const noexcept;
		bool isScalar(node* toCheck) const noexcept;
		void printNodeValueExc(node* nodeElement) const;
		int  printNodeValue(node* nodeElement) const;
	private:
		std::string jsonFilePath;
		pathIndex* fullIndexesList;
		node nodeRoot;
};


}

#endif


include/jsonParserTypes.h
// -----------------------------------------------------------------
// Libjparse - A JSON configuration files parser.
// Copyright (C) 2014  Gabriele Bonacini
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation,
// Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
// -----------------------------------------------------------------

#ifndef __PARSEJSONTYPES__
#define __PARSEJSONTYPES__

#include <stdbool.h>
#include <errno.h>

#define MAX_NODE_STACK      1024
#define MAX_NAME_SIZE       1024
#define MAX_INDEX_SIZE      10240

#define ERR_NO_CONFIG_FILE      2
#define ERR_NO_PARSE            3
#define ERR_WRONG_NODE_TYPE	4
#define ERR_WRONG_INITIAL_NODE	5
#define ERR_UNHANDLED_NODE_TYPE	6
#define ERR_NO_MORE_ELEMENTS	7

#define ERRSTREAM               stderr
#define ROOT_NODE_NAME          "root"
#define DEF_YYWRAP_RETURN       1

typedef enum {MOVE_INNER_OBJECT, MOVE_INNER_ARRAY, DONT_CHANGE_BRANCH, END_OBJ_PENDING, END_OBJ_CLOSED} branch;
typedef enum {ROOT_T, EMPTY_T, NUMERIC_T, TEXT_T, BOOL_T, MATRIX_T, OBJ_T, UNASSIGNED_T} type;

typedef union configData{
	double number;
	char* string;
	bool boolean;
} configData;

typedef struct node node;

struct node{
	char* nodeName;
	type nodeType;
	configData data;
	node* previousNode;
	node* nextNode;
	node* innerNode;
	node* outerNode;
	node* exitNode;
};

typedef struct pathIndex pathIndex;

struct pathIndex{
        char* fullNodeName;
        node* Node;
};

typedef struct pathBuffer pathBuffer;

struct pathBuffer{
        char** nodePathComponent;
        node* Node;
        pathBuffer* Next;
};

void addNode(char* nodeName);
node* allocateEmptyNode(void);
void makeLinks(node* currentNodePtr, node* newNode,type objType);
void initObjName(node* newNode, type objType);
void makePathChain(node* objOrArray);
int createIndex(pathIndex** fullIndexesList);
int indexesListCmp (const void* sbuff1, const void* sbuff2);
int indexesListSearch (const void* sbuff1, const void* sbuff2);
void memoryError(void);
#define checkMemoryErrors(X) if(X == NULL) memoryError()
void spaceError(char* errMsg);
#define chackSpace(X,Y,Z) if(X == Y) spaceError(Z)
/*@null@*/
pathIndex* getElementValueByString(pathIndex* fullIndexesList, char const *key);
bool checkNodeType(node* nodeElement, type expectType);
node* getArrayElements(node* arrayRootElement);
node* getArrayOfObjsElements(node* arrayRootElement);
#define verbose(X,Y) if(X) fprintf(ERRSTREAM, Y);
#define verboseLong(X,Y,Z) if(X) fprintf(ERRSTREAM, Y, Z);
bool isScalar(node* toCheck);
node* getElementFromObj(node* object);

#endif

